src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/roundabout/ComplexRoundaboutFinder.java

```diff
@@ -0,0 +1,45 @@
+package org.openstreetmap.atlas.geography.atlas.items.complex.roundabout;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.openstreetmap.atlas.geography.atlas.Atlas;
+import org.openstreetmap.atlas.geography.atlas.items.Edge;
+import org.openstreetmap.atlas.geography.atlas.items.complex.Finder;
+import org.openstreetmap.atlas.utilities.collections.Iterables;
+
+/**
+ * {@link Finder} for {@link ComplexRoundabout}s.
+ *
+ * @author bbreithaupt
+ */
+public class ComplexRoundaboutFinder implements Finder<ComplexRoundabout>
+{
+    private final Set<Long> checkedIds = new HashSet<>();
+
+    @Override
+    public Iterable<ComplexRoundabout> find(final Atlas atlas)
+    {
+        final List<ComplexRoundabout> complexRoundabouts = new ArrayList<>();
+        Iterables.stream(atlas.edges()).forEach(edge ->
+        {
+            if (!this.checkedIds.contains(edge.getIdentifier()))
+            {
+                final ComplexRoundabout complexRoundabout = new ComplexRoundabout(edge);
+                if (complexRoundabout.getRoundaboutEdgeSet() != null)
+                {
+                    this.checkedIds.addAll(complexRoundabout.getRoundaboutEdgeSet().stream()
+                            .map(Edge::getIdentifier).collect(Collectors.toSet()));
+                }
+                if (complexRoundabout.isValid())
```

Comment:
Is it possible that a roundabout is invalid when starting on Edge X but valid if explored from Edge Y? In that case it might be better to add its edges to the checked IDs only if it is valid?

Reply:
That shouldn't be the case, as the validation of the roundabout is done using the Route built from all the Edges of the roundabout.